### ISSUE TITLE
`useLocalStorage` regression in 7.6.1

### ISSUE THREAD

**Issue by dfaust** [mantinedev/mantine#5848]
### Dependencies check up

- [X] I have verified that I use latest version of all @mantine/* packages

### What version of @mantine/* packages do you have in package.json?

7.6.1

### What package has an issue?

@mantine/hooks

### What framework do you use?

Vite

### In which browsers you can reproduce the issue?

Firefox

### Describe the bug

`useLocalStorage` stores 'undefined' while loading a value.

```ts
  const [settings] = useLocalStorage<Settings>({
    key: 'appSettings',
    getInitialValueInEffect: false,
  });
```

Reverting `@mantine/hooks` to `7.6.0` fixes the problem:

```json
  "dependencies": {
    "@mantine/hooks": "=7.6.0"
  }
```

I assume this is related to #5796.

### Self-service

- [ ] I would be willing to implement a fix for this issue

**Comment by lgaspari** [comment]
Hey @dfaust, how are you doing? Do you think you could provide a `defaultValue` in this case? That way, you'll:

- Get the `defaultValue` as `settings` (instead of `undefined`)
- Store the `defaultValue` on Local Storage under the `appSettings` key

Otherwise, what are you trying to achieve? A codesandbox with a minimal reproduction should help too. Thanks!

**Comment by dfaust** [comment]
Hi @lgaspari,

I don't want to have a default value. I want `useLocalStorage` to return `undefined`, if the value doesn't exist.

If `useLocalStorage` is called and the value doesn't exist, it correctly returns `undefined`. But it also writes a string `"undefined"` to the local storage.
The next time `useLocalStorage` is called, it returns this `"undefined"` string.

![image](https://github.com/mantinedev/mantine/assets/893710/bf33d58f-2c90-41c3-982c-0e4b9d003b71)

**Comment by lgaspari** [comment]
Gotcha, now I understand why this is a **regression** for you, @dfaust.

It might be related to a concern I raised on the linked issue (see https://github.com/mantinedev/mantine/issues/5796#issuecomment-1967944334) where you ALWAYS get the default value on return, therefore, [this side-effect isn't called](https://github.com/mantinedev/mantine/blob/master/packages/%40mantine/hooks/src/use-local-storage/create-storage.ts#L145-L149). In other words, even though my issue was solved, I strongly believe that there are other use cases that aren't considered and might be broken as your case.

I'll give it a try on my end to see if I can help solve your issue while keeping it working for me as well 😄

**Comment by wassim** [comment]
I have the same use case as @dfaust I need the `useLocalStorage` to return null or undefined, not "undefined" as text.

**Comment by lgaspari** [comment]
@dfaust @wassim I've created a fork repository and addressed the issue in there. I'd appreciate if you could give it a try. In order to do so, please follow these steps:

1. Clone the [fork](https://github.com/lgaspari/mantine) repository
2. Install dependencies by using `yarn`
3. Run storybook by using `yarn storybook`
4. Find `use-local-storage` stories (e.g. http://localhost:6006/?path=/story/use-local-storage--no-default-value)

I've covered only 8 scenarios which look similar but they work differently as the following _(not in the same order as stories)_:

| Type | Default value | Storage value | Returned value |
| :-: | :-: | :-: | :-: |
| No SSR | No | No | `undefined` |
| No SSR | No | Yes (`'123'`) | `'123'` |
| No SSR | Yes (`'123'`) | No | `'123'` |
| No SSR | Yes (`'123'`) | Yes (`'654'`) | `'654'` |
| SSR | No | No | `undefined` |
| SSR | No | Yes (`'456'`) | `'456'` |
| SSR | Yes (`'456'`) | No | `'456'` |
| SSR | Yes (`'456'`) | Yes (`'321'`) | `'321'` |

I'll wait for your feedback before creating a pull request :) — if in any case you cannot try it, please let me know.

**Comment by rtivital** [comment]
Fixed in [7.6.2](https://github.com/mantinedev/mantine/releases/tag/7.6.2)

**Comment by dfaust** [comment]
@lgaspari: I didn't try your fix, but @rtivital's fix in 7.6.2 works.

Thanks everybody!
